cli/src/lib.rs

```diff
@@ -83,6 +84,20 @@ where
     pub block_sync: AlwaysAddr<B>,
     /// Torii web server
     pub torii: Option<Torii>,
+    /// Thread handlers
+    thread_handlers: Vec<ThreadHandler>,
+}
+
+impl<G, S, B> Drop for Iroha<G, S, B>
+where
+    G: GenesisNetworkTrait,
+    S: SumeragiTrait<GenesisNetwork = G>,
+    B: BlockSynchronizerTrait<Sumeragi = S>,
+{
+    fn drop(&mut self) {
+        // Drop thread handles first
+        let _thread_handles = core::mem::take(&mut self.thread_handlers);
```

Comment:
You can remove this explicit earlier drop and just place `thread_handlers` at the top of the `Iroha` structure. See order of drops [here](https://doc.rust-lang.org/reference/destructors.html).

But I'dd add a description comment, that noone should place something higher than `thread_handlers`.

Both options with *Please don't modify* comment and `mem::take()` looks like a crutch for me. Not sure, which one is better.

Reply:
You seem to assume *implicit* is somehow inherently better than being *explicit*. If anything I'd argue that the opposite is true.

Code that is *explicit* is **less clever**. Code that is *less clever* is **easier to understand**. Code that is *easier to understand* is more **debuggable** and more **maintainable**.

I think the use of `mem::take` is very tasteful and nothing ought to be changed.

Reply:
I read through issue regarding stabilization of drop order and it seems like it was created because too many users of language relied on unspecified behavior. 
So i want to be explicit on this one and decide to implement `Drop` manually.

Reply:
@SamHSmith, just my note about *implicit* and *explicit*.

It's not an absolute. We do not choose between *all should be implicit* and *all should be explicit*. There is a balance.

If you want everything to be explicit -- there is the assembly language. But we do not operate with it. It's **too** explicit for our purposes. We operate with some abstractions.

Same is here. There are a lot of abstractions and agreements in **Rust**. One agreement is -- the order of drops in structure is the same as the order of field declarations. This is how strucutres in Rust work. If you are working with Rust you should know it. But with `core::mem::take()` we are breaking this agreement. We are breaking the agreement of the tool we use. The agreement that everyone expects looking at structure declaration. We are not using things that were designed for that, we are doing it manually (which can be errorprone sometimes; less code -- better code).

We are placing `Default` value into the field and then let Rust drop it in the defined order. That's why I think it's a **crutch**. 

This `Drop` implementation is hard to understand without the helpful comment. Without that comment this implementation makes a lot of questions like *Why? For what do we call `take()` here?*

Although I don't insist on my approach. Just wanted to respond to you